### ISSUE TITLE
Generated cgo rule includes platform-specific clinkopts

### ISSUE THREAD

**Issue by kgreenek** [bazelbuild/rules_go#402]
I need github.com/mattn/go-sqlite3 as a dependency in my Go project. I am adding it via a new_go_repository() rule in my WORKSPACE file.

Internally, it has a file called sqlite3_windows.go that contains the following content:

```go
// Copyright (C) 2014 Yasuhiro Matsumoto <mattn.jp@gmail.com>.
//
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file.
// +build windows

package sqlite3

/*
#cgo CFLAGS: -I. -fno-stack-check -fno-stack-protector -mno-stack-arg-probe
#cgo windows,386 CFLAGS: -D_USE_32BIT_TIME_T
#cgo LDFLAGS: -lmingwex -lmingw32
*/
import "C"
```

The generated BUILD.bazel file contains the following rule:

```
cgo_library(
    name = "cgo_default_library",
    srcs = [
        "backup.go",
        "callback.go",
        "error.go",
        "sqlite3.go",
        "sqlite3-binding.c",
        "sqlite3-binding.h",
        "sqlite3_context.go",
        "sqlite3_fts5.go",
        "sqlite3_icu.go",
        "sqlite3_json1.go",
        "sqlite3_libsqlite3.go",
        "sqlite3_load_extension.go",
        "sqlite3_omit_load_extension.go",
        "sqlite3_other.go",
        "sqlite3_trace.go",
        "sqlite3_type.go",
        "sqlite3_vtable.go",
        "sqlite3_windows.go",
        "sqlite3ext.h",
    ],
    clinkopts = [
        "-lm",
        "-licuuc",
        "-licui18n",
        "-lsqlite3",
        "-ldl",
        "-lmingwex",
        "-lmingw32",
    ],
    copts = [
        "-std=gnu99",
        "-DSQLITE_ENABLE_RTREE",
        "-DSQLITE_THREADSAFE",
        "-DSQLITE_ENABLE_FTS3",
        "-DSQLITE_ENABLE_FTS3_PARENTHESIS",
        "-DSQLITE_ENABLE_FTS4_UNICODE61",
        "-DSQLITE_TRACE_SIZE_LIMIT=15",
        "-DSQLITE_DISABLE_INTRINSIC",
        "-Wno-deprecated-declarations",
        "-DSQLITE_ENABLE_FTS5",
        "-DSQLITE_ENABLE_ICU",
        "-DSQLITE_ENABLE_JSON1",
        "-DUSE_LIBSQLITE3",
        "-DSQLITE_OMIT_LOAD_EXTENSION",
        "-I.",
        "-std=gnu99",
        "-DSQLITE_ENABLE_RTREE",
        "-DSQLITE_THREADSAFE",
        "-DSQLITE_ENABLE_FTS3",
        "-DSQLITE_ENABLE_FTS3_PARENTHESIS",
        "-DSQLITE_ENABLE_FTS4_UNICODE61",
        "-DSQLITE_TRACE_SIZE_LIMIT=15",
        "-DSQLITE_ENABLE_COLUMN_METADATA=1",
        "-Wno-deprecated-declarations",
        "-I.",
        "-fno-stack-check",
        "-fno-stack-protector",
        "-mno-stack-arg-probe",
    ],
    visibility = ["//visibility:private"],
    deps = ["@org_golang_x_net//context:go_default_library"],
)
```

My build fails to link because mingex and mingw32 libraries cannot be found. However, I'm building on Ubuntu, so that shouldn't be a problem according to the "+build windows" rule at the top of sqlite3_windows.go.

**Comment by kgreenek** [comment]
As a work-around, I cloned the project and deleted the sqlite3_windows.go file, since I don't need to support windows.

**Comment by jayconrod** [comment]
I don't think there's an easy solution to this problem. We don't want Gazelle to generate platform-specific BUILD files, which means it shouldn't apply build constraints. In this situation though, it creates dependencies on system libraries that don't exist, which is clearly not an ideal outcome.

Some ideas:

* We could make the rules responsible for extracting flags from cgo flags at compile time after applying build constraints. This would simulate what "go build" does. There's no way to do in our current implementation of cgo though, since we pass the flags to a separate `cc_library` rule, generated in advance.
* We could generate `config_setting` rules and use `select` to pick the right set of flags for each platform. In general, I think it's very difficult to map from build tags to `config_setting` rules, and I'm not sure we could get tools to do this reliably.

**Comment by jayconrod** [comment]
As a more stable workaround, use the `build_tags` attribute in `new_go_repository`. Put all tags in the list you want to evaluate as true.

Example:

```bzl
new_go_repository(
    name = "com_github_mattn_go_sqlite3",
    importpath = "github.com/mattn/go-sqlite3",
    tag = "v1.2.0",
    build_tags = ["linux","amd64"],
)
```

**Comment by kgreenek** [comment]
Thanks! That's a nicer fix for this situation than forking and deleting the windows file.

I think the idea in your first bullet point makes a ton of sense. That is how I would expect it to work. And flags that you add in the BUILD file would over-ride flags that the go rule auto-detected.

**Comment by jayconrod** [comment]
This specific issue has been fixed in #509. However, `github.com/mattn/go-sqlite3` still doesn't build as an external dependency due to a cgo issue. See #526.
